pkg/retainer/retainer/retainer.go

```diff
@@ -0,0 +1,294 @@
+	return module.Provide(func(e *engine.Engine) retainer.Retainer {
+		r := New(e.Storage.Retainer,
+			func(blockID iotago.BlockID) *model.Block {
+				block, exists := e.BlockFromCache(blockID)
+				if !exists {
+					return nil
+				}
+
+				return block.ModelBlock()
+			},
+			func(blockID iotago.BlockID) (*model.Block, error) {
+				storage := e.Storage.Blocks(blockID.Index())
+				if storage == nil {
+					return nil, ierrors.Errorf("failed to get storage for block %s in retainer", blockID)
+				}
+
+				block, err := e.Storage.Blocks(blockID.Index()).Load(blockID)
+				if err != nil {
+					return nil, ierrors.Wrap(err, "failed to get block from storage in retainer")
+				}
+
+				if block == nil {
+					return nil, ierrors.Errorf("failed to get block from storage in retainer, block not found: %s", blockID)
+				}
+
+				return block, nil
+			}, e.Storage.Settings().LatestFinalizedSlot,
+			e.Storage.Settings().LatestCommitment().Index,
+			e.ErrorHandler("retainer"))
+
+		asyncOpt := event.WithWorkerPool(e.Workers.CreatePool("Retainer", 1))
+
+		e.Events.BlockDAG.BlockAttached.Hook(func(b *blocks.Block) {
+			_, isTx := b.Transaction()
+			if err := r.onBlockAttached(b.ID(), isTx); err != nil {
+				r.errorHandler(ierrors.Wrap(err, "failed to store on BlockAttached in retainer"))
+			}
+		}, asyncOpt)
+
+		e.Events.BlockGadget.BlockAccepted.Hook(func(b *blocks.Block) {
+			if err := r.onBlockAccepted(b.ID()); err != nil {
+				r.errorHandler(ierrors.Wrap(err, "failed to store on BlockAccepted in retainer"))
+			}
+		}, asyncOpt)
+
+		e.Events.BlockGadget.BlockConfirmed.Hook(func(b *blocks.Block) {
+			if err := r.onBlockConfirmed(b.ID()); err != nil {
+				r.errorHandler(ierrors.Wrap(err, "failed to store on BlockConfirmed in retainer"))
+			}
+		}, asyncOpt)
+
+		e.HookInitialized(func() {
+			e.Ledger.OnTransactionAttached(func(transactionMetadata mempool.TransactionMetadata) {
+				attachmentID := transactionMetadata.EarliestIncludedAttachment()
+
+				if err := r.onTransactionAttached(attachmentID); err != nil {
+					r.errorHandler(ierrors.Wrap(err, "failed to store on TransactionAttached in retainer"))
+				}
+
+				transactionMetadata.OnAccepted(func() {
+					if slotIndex := attachmentID.Index(); slotIndex > 0 {
+						if err := r.onTransactionAccepted(attachmentID); err != nil {
+							r.errorHandler(ierrors.Wrap(err, "failed to store on TransactionAccepted in retainer"))
+						}
+					}
+				})
+
+				transactionMetadata.OnEarliestIncludedAttachmentUpdated(func(prevBlock, newBlock iotago.BlockID) {
+					if newBlock.Index() != prevBlock.Index() {
+						if err := r.onAttachmentUpdated(prevBlock, newBlock); err != nil {
+							r.errorHandler(ierrors.Wrap(err, "failed to delete/store on AttachmentUpdated in retainer"))
+						}
+					}
+				})
+			})
+		})
+
+		r.TriggerInitialized()
+
+		return r
+	})
+}
+
+func (r *Retainer) Shutdown() {
+	r.retainerFunc = nil
```

Comment:
this is dangerous because if any of the on* still gets called it will crash with a nil pointer exception